lib/spack/spack/binary_distribution.py

```
@@ -536,30 +824,50 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
     return None
 
 
-def download_tarball(spec):
+def download_tarball(spec, url=None):
     """
     Download binary tarball for given package into stage area
     Return True if successful
+
+    Args:
+        spec (Spec): Concrete spec
+        url (str): If provided, this is the preferred url, other configured
+            mirrors will only be used if the tarball can't be retrieved from
+            this preferred url.
+
+    Returns:
+        Path to the downloaded tarball, or ``None`` if the tarball could not
+            be downloaded from any configured mirrors.
     """
     if not spack.mirror.MirrorCollection():
         tty.die("Please add a spack mirror to allow " +
                 "download of pre-compiled packages.")
 
     tarball = tarball_path_name(spec, '.spack')
 
+    urls_to_try = []
+
+    if url:
+        urls_to_try.append(url_util.join(
+            url, _build_cache_relative_path, tarball))
+
     for mirror in spack.mirror.MirrorCollection().values():
-        url = url_util.join(
-            mirror.fetch_url, _build_cache_relative_path, tarball)
+        if not url or url != mirror.fetch_url:
+            urls_to_try.append(url_util.join(
+                mirror.fetch_url, _build_cache_relative_path, tarball))
 
+    for try_url in urls_to_try:
         # stage the tarball into standard place
-        stage = Stage(url, name="build_cache", keep=True)
+        stage = Stage(try_url, name="build_cache", keep=True)
         stage.create()
         try:
             stage.fetch()
             return stage.save_filename
         except fs.FetchError:
             continue
 
+    # If we arrive here, something went wrong, maybe this should be an
+    # exception?
```

Could you elaborate? I think the issue is that if we are in this function, we have established that a binary cache claims to have a spec, but if we get to this point, we weren't actually able to download it. It may be worth printing a warning but as far as I can tell this was possible before this PR, so I don't think it should raise an error.

For example I think this could happen if there is an out-of-date index.json on the remote binary cache and the maintainer of that cache removed the file - which isn't an error on our end (I think the install should continue from source in that case).

---

> I think the issue is that if we are in this function, we have established that a binary cache claims to have a spec, but if we get to this point, we weren't actually able to download it.

Yes, that's the idea, no matter how you got in here, we should have already verified that the spec exists on one of the mirrors.  I agree that installing from source when it's missing is the right thing to do.  I'll just print a warning though, as you suggest.